tests/test_containers.py

```diff
@@ -0,0 +1,200 @@
+
+
+logger = logging.getLogger(__file__)
+
+
+@pytest.fixture
+def dcli():
+    """Docker client object."""
+    return docker.from_env()
+
+
+@pytest.fixture
+def docker_dir(test_dir):
+    """Path to the tests/docker directory."""
+    return test_dir / 'docker'
+
+
+def _build_image(dcli, docker_dir, name):
+    img, _ = dcli.images.build(
+            path=str(docker_dir),
+            tag=f'mahiru-test/{name}',
+            dockerfile=str(docker_dir / f'{name}.Dockerfile'),
+            rm=True)
+    return img
+
+
+def _make_tarfile(dcli, docker_dir, image, name):
+    tar_path = docker_dir / f'{name}.tar.gz'
+    with gzip.open(str(tar_path), 'wb', compresslevel=1) as f:
+        for chunk in image.save():
+            f.write(chunk)
+    return tar_path
+
+
+def _build_data_images(dcli, docker_dir):
+    base_image = None
+    input_image = None
+    base_file = None
+    input_file = None
+    try:
+        base_image = _build_image(dcli, docker_dir, 'data-asset-base')
+        input_image = _build_image(dcli, docker_dir, 'data-asset-input')
+        base_file = _make_tarfile(
+                dcli, docker_dir, base_image, 'data-asset-base')
+        input_file = _make_tarfile(
+                dcli, docker_dir, input_image, 'data-asset-input')
+        return base_file, input_file
+    except Exception:
+        if input_file:
+            input_file.unlink(missing_ok=True)
+        if base_file:
+            base_file.unlink(missing_ok=True)
+        raise
+    finally:
+        if input_image:
+            dcli.images.remove('mahiru-test/data-asset-input')
+        if base_image:
+            dcli.images.remove('mahiru-test/data-asset-base')
+
+
+def _build_compute_image(dcli, docker_dir):
+    base_image = None
+    input_image = None
+    input_file = None
+    try:
+        base_image = _build_image(dcli, docker_dir, 'compute-asset-base')
+        input_image = _build_image(dcli, docker_dir, 'compute-asset')
+        input_file = _make_tarfile(
+                dcli, docker_dir, input_image, 'compute-asset')
+        return input_file
+    except Exception:
+        if input_file:
+            input_file.unlink(missing_ok=True)
+        raise
+    finally:
+        if input_image:
+            dcli.images.remove('mahiru-test/compute-asset')
+        if base_image:
+            dcli.images.remove('mahiru-test/compute-asset-base')
+
+
+@pytest.fixture
+def data_asset_tars(dcli, docker_dir):
+    base_file, input_file = _build_data_images(dcli, docker_dir)
+    yield base_file, input_file
+    input_file.unlink()
+    base_file.unlink()
+
+
+@pytest.fixture
+def compute_asset_tar(dcli, docker_dir):
+    compute_file = _build_compute_image(dcli, docker_dir)
+    yield compute_file
+    compute_file.unlink()
+
+
+@patch(
+        'proof_of_concept.components.domain_administrator.OUTPUT_ASSET_ID',
+        Identifier('asset:ns:output_base:ns:test_site'))
+def test_container_step(
```

Comment:
Really cool that this works!

Reply:
Yeah, although it's a bit of a hack. Eventually the compute asset should have some metadata that specifies which base image to use for each output, so that you can have outputs as files (on a  WebDAV) or as triples (in a Virtuoso server or something) or records (in a relational database). And then type declarations for inputs and outputs so that you can check that your workflow ties things together correctly, and ... but we'll probably never get there, and this works.